core/solver/batch_bicgstab_kernels.hpp

```diff
@@ -0,0 +1,122 @@
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_SOLVER_BATCH_BICGSTAB_KERNELS_HPP_
+#define GKO_CORE_SOLVER_BATCH_BICGSTAB_KERNELS_HPP_
+
+
+#include <ginkgo/core/base/batch_multi_vector.hpp>
+#include <ginkgo/core/log/batch_logger.hpp>
+#include <ginkgo/core/matrix/batch_dense.hpp>
+#include <ginkgo/core/matrix/batch_ell.hpp>
+#include <ginkgo/core/stop/batch_stop_enum.hpp>
+
+
+#include "core/base/kernel_declaration.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace batch_bicgstab {
+
+
+/**
+ * Options controlling the batch Bicgstab solver.
+ */
+template <typename RealType>
+struct BicgstabSettings {
+    int max_iterations;
+    RealType residual_tol;
+    ::gko::batch::stop::ToleranceType tol_type;
+};
+
+
+/**
+ * Calculates the amount of in-solver storage needed by batch-Bicgstab.
+ *
+ * The calculation includes multivectors for
+ * - r
+ * - r_hat
+ * - p
+ * - p_hat
+ * - v
+ * - s
+ * - s_hat
+ * - t
+ * - x
+ * Note: small arrays for
+ * - rho_old
+ * - rho_new
+ * - omega
+ * - alpha
+ * - temp
+ * - rhs_norms
+ * - res_norms
+ * are currently not accounted for as they are in static shared memory.
+ */
+template <typename ValueType>
+inline int local_memory_requirement(const int num_rows, const int num_rhs)
+{
+    return (9 * num_rows * num_rhs) * sizeof(ValueType);
+}
+
+
+}  // namespace batch_bicgstab
+
+
+#define GKO_DECLARE_BATCH_BICGSTAB_APPLY_KERNEL(_type)                       \
+    void apply(                                                              \
+        std::shared_ptr<const DefaultExecutor> exec,                         \
+        const gko::kernels::batch_bicgstab::BicgstabSettings<                \
```

Comment:
All Ginkgo kernels are in `gko::kernels` namespace. I found it much cleaner to have the batch kernels in `batch_x` instead of nesting them in `batch::x`.